src/recordlinker/linking/link.py

```diff
@@ -113,36 +114,58 @@ def link_record_against_mpi(
+                            "algorithm.ratio_upper": belongingness_ratio_upper_bound,
+                            "algorithm.pass.blockingkeys": algorithm_pass.blocking_keys,
+                            "algorithm.pass.evaluators": algorithm_pass.evaluator_features,
+                        },
+                    )
                     if belongingness_ratio >= belongingness_ratio_lower_bound:
                         # The match ratio is larger than the minimum cluster threshold,
                         # optionally update the max score for this person
                         scores[person] = max(scores[person], belongingness_ratio)
 
+    prediction: schemas.Prediction = "possible_match"
     matched_person: typing.Optional[models.Person] = None
-    if scores:
-        # Find the person with the highest matching score
-        matched_person, _ = max(scores.items(), key=lambda i: i[1])
-
-    sorted_scores: list[LinkResult] = [LinkResult(k, v) for k, v in sorted(scores.items(), reverse=True, key=lambda item: item[1])]
-    if not scores:
+    results: list[LinkResult] = [
+        LinkResult(k, v) for k, v in sorted(scores.items(), reverse=True, key=lambda i: i[1])
+    ]
+    if not results:
         # No match
-        matched_person = models.Person() # Create new Person Cluster
-        results = []
-    elif sorted_scores[0].belongingness_ratio >= belongingness_ratio_upper_bound:
+        prediction = "no_match"
+        matched_person = models.Person()  # Create new Person Cluster
+    elif results[0].belongingness_ratio >= belongingness_ratio_upper_bound:
         # Match (1 or many)
-        matched_person = sorted_scores[0].person
-        results = [x for x in sorted_scores if x.belongingness_ratio >= belongingness_ratio_upper_bound] # Multiple matches
+        prediction = "match"
+        matched_person = results[0].person
+        # reduce results to only those that meet the upper bound threshold
+        results = [x for x in results if x.belongingness_ratio >= belongingness_ratio_upper_bound]
         if not algorithm.include_multiple_matches:
-            results = [results[0]] # 1 Match (highest Belongingness Ratio)
-    else:
-        # Possible match
-        matched_person = None
-        results = sorted_scores
+            # reduce results to only the highest match
+            results = [results[0]]
 
     with TRACER.start_as_current_span("insert"):
         patient = mpi_service.insert_patient(
             session, record, matched_person, record.external_id, external_person_id, commit=False
         )
 
+    LOGGER.info(
+        "link results",
+        extra={
+            "person.reference_id": matched_person and matched_person.reference_id,
+            "patient.reference_id": patient.reference_id,
+            "result.prediction": prediction,
+            "result.count": len(results),
+            "result.scored_count": len(scores),
```

Comment:
What is the purpose of logging both `len(results)` and `len(scores)`?

Reply:
`len(results)` is the number of results we return to the user, so its showing us how often we have more than 1 record in the upper threshold. Although as I write this, thats only true if the `include_multiple_matches == true` , so maybe I need to revise this to work in all cases.  My thought is that if we often get results where this number is above 1, clustering is suboptimal. 

`len(scores)` is telling us how many results are above the lower threshold.  I was thinking this would somehow convey to us how well the blocking queries are being set up (or are we too limiting, or are we including too many records), but again there may be a mistake here in that we aren't including results that are below the threshold.  So its not exact, but maybe a proxy for how well the blocking queries are performing?

Thanks for questioning this, we can probably to better with these metrics.

Reply:
>My thought is that if we often get results where this number is above 1, clustering is suboptimal.

I think this depends on what the upper threshold is (in addition to whether `include_multiple_matches=true`, as you said). I'm not sure if this is a valid interpretation in all cases. For example: 

1. If `belongingness_ratio=(0.9, 0.92)` and `include_multiple_matches=true`, then that might be a fair assumption (interpretation of `len(results)` > 1: multiple matches were found where >= 90% Patient records match the incoming Patient record - should these two clusters be the same Person? Can point to poor-quality clusters)
2. If `belongingness_ratio=(0.9, 0.92)` and `include_multiple_matches=false`, `len(results)` will always be 1, which is not meaningful information.
3. If `belongingness_ratio=(0.75, 0.9)` and `include_multiple_matches=true`, then I'm not sure this necessarily points to poor clusters. This would just mean that all results where >= 75% Patient records match the incoming Patient record are returned, and 75% can be a relatively low threshold for Belongingness. We would need to understand if there are Belongingness Ratios that tend higher to the 90% within the `results` for the assumption that clustering is suboptimal to be valid, I think. This would require stratification that doesn't seem feasible or too helpful right now.
4. If `belongingness_ratio=(0.75, 0.9)` and `include_multiple_matches=false` same as 2.

Reply:
Either way, I think it would be helpful to include in the metrics 1) the number of results above the upper threshold, and 2) the number of results above the lower threshold. The statistical interpretations will vary, depending on `belongingness_ratio` and `include_multuple_matches`, as discussed above. But I think we'd just need to distinguish what these metrics are more clearly than `len(results)` and `len(scores)`. Maybe `matches_above_lower_bound` and `matches_above_upper_bound`?

Reply:
Yep, I've reimplemented how these are calculated.  Take a look at the new `result_counts` variable.